tests/weak_forms/step-6-variant_02-energy_functional_sd_04.cc

```diff
@@ -0,0 +1,163 @@
+//
+// This file is part of the Weak forms for deal.II library.
+//
+// The Weak forms for deal.II library is free software; you can use it,
+// redistribute it, and/or modify it under the terms of the GNU Lesser
+// General Public License as published by the Free Software Foundation;
+// either version 3.0 of the License, or (at your option) any later
+// version. The full text of the license can be found in the file LICENSE
+// at the top level of the Weak forms for deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+// Laplace problem: Assembly using self-linearizing energy functional weak form
+// in conjunction with symbolic differentiation.
+// The energy functional form is recovered directly from a symbolic function.
+// This test replicates step-6 exactly.
+// - Optimizer type: LLVM
+// - Optimization method: All
+
+#include <weak_forms/weak_forms.h>
+
+#include "../weak_forms_tests.h"
+#include "wf_common_tests/step-6.h"
+
+
+using namespace dealii;
+
+
+template <int dim>
+class Step6 : public Step6_Base<dim>
+{
+public:
+  Step6();
+
+protected:
+  void
+  assemble_system() override;
+};
+
+
+template <int dim>
+Step6<dim>::Step6()
+  : Step6_Base<dim>()
+{}
+
+
+template <int dim>
+void
+Step6<dim>::assemble_system()
+{
+  using namespace WeakForms;
+  using namespace Differentiation;
+
+  constexpr int spacedim = dim;
+  using SDNumber_t       = Differentiation::SD::Expression;
+
+  // Symbolic types for test function, trial solution and a coefficient.
+  const TestFunction<dim>  test;
+  const FieldSolution<dim> solution;
+  const ScalarFunctor      rhs_coeff("s", "s");
+
+  const WeakForms::SubSpaceExtractors::Scalar subspace_extractor(0, "s", "s");
+
+  const auto test_ss  = test[subspace_extractor];
+  const auto test_val = test_ss.value();
+
+  const auto soln_ss   = solution[subspace_extractor];
+  const auto soln_grad = soln_ss.gradient();
+
+  const auto coefficient = ScalarFunctor("c1", "c1")
+                             .template value<double, dim, spacedim>(
+                               [](const FEValuesBase<dim, spacedim> &fe_values,
+                                  const unsigned int                 q_point)
+                               {
+                                 const Point<spacedim> &p =
+                                   fe_values.get_quadrature_points()[q_point];
+                                 return (p.square() < 0.5 * 0.5 ? 20.0 : 1.0);
+                               },
+                               UpdateFlags::update_quadrature_points);
+
+  const auto energy = 0.5 * coefficient * scalar_product(soln_grad, soln_grad);
+
+  const auto rhs_coeff_func = rhs_coeff.template value<double, dim, spacedim>(
+    [](const FEValuesBase<dim, spacedim> &, const unsigned int)
+    { return 1.0; });
+
+  MatrixBasedAssembler<dim> assembler;
+  assembler += energy_functional_form<dim, spacedim>("e", "\\Psi", energy).dV();
```

Comment:
@Rombur, this might interest you: This enhancement that I'm working on allows one to write a composite expression, and pass it directly into a self-linearising form. The linearisation is then done on two levels automatically. First the energy form is converted into the linear and bilinear forms, as was always done by the `energy_functional` (and similarly the `residual_view`) forms. But then the low level differentiation is taken care of in the background, and any of the complexity that would normally appear by using the SD wrappers is completely hidden from view.

Reply:
That looks great.  Thanks for letting me know.